platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -62,30 +73,174 @@ public class JUnitTestEventAdapter extends RunListener {
         new AssertErrorMapper()
     );
 
+    private static final class TestNode {
+        // Avoid generating the id too early, as it leads to a confusing ordering of ids
+        // Ordering generally shouldn't be relied on, but the tests read cleaner if ids are generated in a depth-first order
+        private final IdGenerator<?> idGenerator;
+        @Nullable
+        private volatile Object resolvedId;
+        final Description description;
+
+        TestNode(IdGenerator<?> idGenerator, Description description) {
+            this.idGenerator = idGenerator;
+            this.description = description;
+        }
+
+        public Object resolveId() {
+            Object localId = resolvedId;
+            if (localId == null) {
+                synchronized (this) {
+                    localId = resolvedId;
+                    if (localId == null) {
+                        localId = idGenerator.generateId();
+                        resolvedId = localId;
+                    }
+                }
+            }
+            return localId;
+        }
+    }
+
+    private static final class PostRunStartData {
+        final Map<Description, TestNode> parentDescToNode;
+        final Map<Description, TestNode> childDescToParent;
+
+        PostRunStartData(
+            Map<Description, TestNode> parentDescToNode,
+            Map<Description, TestNode> childDescToParent
+        ) {
+            this.parentDescToNode = parentDescToNode;
+            this.childDescToParent = childDescToParent;
+        }
+    }
+
     private static final DefaultThrowableToTestFailureMapper FAILURE_MAPPER = new DefaultThrowableToTestFailureMapper(MAPPERS);
 
     private static final Pattern DESCRIPTOR_PATTERN = Pattern.compile("(.*)\\((.*)\\)(\\[\\d+])?", Pattern.DOTALL);
     private final IdGenerator<?> idGenerator;
     private final TestResultProcessor resultProcessor;
     private final Clock clock;
     private final Object lock = new Object();
-    private final Map<Description, TestDescriptorInternal> executing = new HashMap<Description, TestDescriptorInternal>();
-    private final Set<Description> assumptionFailed = new HashSet<Description>();
+    @Nullable
+    private volatile PostRunStartData postRunStartData;
+    // This uses a Deque so grandparents are completed after parents
+    private final Deque<Description> activeParents = new ConcurrentLinkedDeque<>();
+    private final Map<Description, TestDescriptorInternal> executing = new HashMap<>();
+    private final Set<Description> assumptionFailed = new HashSet<>();
+    private volatile boolean testsStarted = false;
+    @Nullable
+    private volatile String rootClassName;
```

Comment:
💅 The root might not be a class soon, can we rename this?
```suggestion
    private volatile String rootNodeName;
```